gh-143304: Fix ctypes.CDLL to honor handle parameter on POSIX systems

Koolvansh07 opened the pull request:
The `handle` parameter was being ignored in the POSIX implementation of `CDLL._load_library()`, causing it to always call `_dlopen()` even when a valid handle was provided. This was a regression introduced in recent refactoring.

This PR adds the missing handle check to match the Windows implementation behavior, and includes a regression test.

Fixes gh-143304

Koolvansh07 committed:
Fix ctypes.CDLL to honor handle parameter on POSIX systems

The handle parameter was being ignored in the POSIX implementation
of CDLL._load_library(), causing it to always call _dlopen() even
when a valid handle was provided. This was a regression introduced
in recent refactoring.

This commit adds the missing handle check to match the Windows
implementation behavior, and includes a regression test.

Fixes gh-143304

Koolvansh07 committed:
Add NEWS entry for gh-143304

picnixz commented on Lib/test/test_ctypes/test_loading.py:
```diff
@@ -106,6 +106,20 @@ def test_load_without_name_and_with_handle(self):
+    def test_load_without_name_and_with_handle_posix(self):
+        # Test that CDLL honors the handle parameter on POSIX systems
+        # This is a regression test for gh-143304
+        if libc_name is None:
```
Add a skipIf instead.

picnixz commented on Misc/NEWS.d/next/Library/2026-01-01-05-26-00.gh-issue-143304.Kv7x9Q.rst:
```diff
@@ -0,0 +1 @@
+Fix :class:`ctypes.CDLL` to honor the ``handle`` parameter on POSIX systems. The parameter was being ignored, causing the library to always call ``_dlopen()`` even when a valid handle was provided.
```
Don't mention `dlopen`, it's private.

picnixz commented on Lib/test/test_ctypes/test_loading.py:
```diff
@@ -106,6 +106,20 @@ def test_load_without_name_and_with_handle(self):
+        # This is a regression test for gh-143304
+        if libc_name is None:
+            self.skipTest('could not find libc')
+        # First load a library normally to get a handle
```
Remove trivial comments. Was this generated by an LLM? Please read https://devguide.python.org/getting-started/generative-ai/.

Koolvansh07 replied:
Alright making these changes asap and I'll follow these guidelines thank you

picnixz replied:
Please answer the question as well: was it LLM generated? or LLM-aided?

picnixz commented on Lib/ctypes/__init__.py:
```diff
@@ -458,6 +458,8 @@ def _load_library(self, name, mode, handle, winmode):
                 if name and name.endswith(")") and ".a(" in name:
                     mode |= _os.RTLD_MEMBER | _os.RTLD_NOW
             self._name = name
+            if handle is not None:
+                return handle
```
No need for computing `mode` in this case I think?

Koolvansh07 committed:
Address review feedback

- Remove AI-generated comments from test
- Use skipIf decorator instead of runtime check
- Simplify NEWS entry (don't mention private _dlopen)

Koolvansh07 committed:
Move handle check to beginning to avoid unnecessary computation

picnixz commented on test_handle_fix.py:
What are those files?

Koolvansh07 replied:
for testing, accidentally pushed them getting rid of them

picnixz replied:
It's also AI-generated.

Koolvansh07 replied:
Sorry about that! Those were debug files I created while testing the fix locally. I've removed them in the latest commit.

Koolvansh07 committed:
Remove debug files that were accidentally committed

picnixz commented on Lib/ctypes/__init__.py:
```diff
@@ -440,6 +440,9 @@ def _load_library(self, name, mode, handle, winmode):
         def _load_library(self, name, mode, handle, winmode):
+            if handle is not None:
+                self._name = name
+                return handle
```
Ah I'm sorry here, I didn't see that `name`may change. So let's put this at the end of the function like for the Windows case (just before the dlopen call). That's my bad.

Koolvansh07 replied:
no problem 🫡

picnixz commented on Lib/test/test_ctypes/test_loading.py:
```diff
@@ -106,6 +106,14 @@ def test_load_without_name_and_with_handle(self):
+        lib1 = CDLL(libc_name)
+        handle = lib1._handle
+        lib2 = CDLL(name=None, handle=handle)
+        self.assertIs(handle, lib2._handle)
```
```suggestion
        self.assertIs(lib2._handle, handle)
```

Can you check whether there were tests for Python 3.12 where we had this handle? or if the tests were rewritten as well (see the PRs mentioned on the issue).

Koolvansh07 replied:
I checked the `3.12` branch source code:

1. `Lib/ctypes/__init__.py` in 3.12 DID support the `handle` parameter (it checked `if handle is None:`).
2. 2. `Lib/test/test_ctypes/test_loading.py` in 3.12 DID NOT have any test using `CDLL(..., handle=...)` on POSIX.

picnixz replied:
Ok thx!

Koolvansh07 committed:
Move handle check to end of function as suggested by reviewer

The name parameter may be modified by .fwork and AIX processing,
so we need to process it before checking the handle.

Koolvansh07 committed:
Update test assertion to match (actual, expected) convention

sharktide reviewed (approved):
LGTM
Thanks for your contribution @Koolvansh07

serhiy-storchaka reviewed (approved):
I thought it was Windows-only parameter, but it is documented on Posix too.

LGTM. :+1:
